Added Dockerfile, .dockerignore

Ts1n opened the pull request:
Closes #12

Ts1n committed:
Added Dockerfile, .dockerignore

Ts1n committed:
Merge branch 'master' into feature/12

pmhsfelix commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+WORKDIR /app
+
+# Copy generated .jar
+COPY --from=build-env /src/build/libs/*.jar i-on-integration.jar
```
See https://github.com/i-on-project/example-ci-cd/blob/experiment0/Dockerfile for a layered way of building the docker file, avoiding creating a single layer containing both the dependent libs (which can be large, but don't change often) and app classes (which are small and change often).

Ts1n replied:
That's a better solution. I've change the Dockerfile.
In the example shown I don't understand the need of creating a new user...

pmhsfelix replied:
Increased security. https://spring.io/guides/gs/spring-boot-docker/ and https://security.stackexchange.com/questions/106860/can-a-root-user-inside-a-docker-lxc-break-the-security-of-the-whole-system

Ts1n committed:
Merge branch 'master' into feature/12

pmhsfelix commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+WORKDIR /src
+
+# Copy src .dockerignore prevents copy of unnecessary files
+COPY . .
```
Does this only copy the `src` folder? We also need files in the root folder and other nested folders (e.g. checkstyle) to perform the build.

Ts1n replied:
This copies the whole `root folder` of the project (`src folder`, `gradle` and others) ignoring the files mention on `.dockerignore` to the current folder (`WORKDIR`) of the container

pmhsfelix commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+COPY . .
+
+# Gradle build with daemon disabled
+RUN ./gradlew build --no-daemon --stacktrace
```
What do we gain by doing the build inside docker?

Ts1n replied:
We remove the need for the build machine to have the necessary build tools. I believe it to be more portable.

pmhsfelix reviewed (commented):
Waiting for answers to questions, before approving.

Ts1n committed:
Merge branch 'master' into feature/12

Ts1n committed:
Multiple changes

$Added gradle task to extract uber Jar; Fixed Dockerfile to take advantage of image layers

Ts1n committed:
Added spring user to avoid running app as root
